rfcs/document_all_media_properties/20230307-project_proposal.md

```diff
@@ -0,0 +1,133 @@
+
+## Due date:
+
+2023-03-21
+
+## Assigned reviewers
+
+- [ ] @AetherUnbound
+- [ ] @krysal
+
+## Description
+
+### Year Goal
+
+Overcoming **Data Inertia**.
+
+### Motivation
+
+The first step in making the data we have reliable is describing it in detail.
+Currently, we have more than 600 million rows of data about media in the catalog
+database and similar but slightly different sets of data in the API database and
+elasticsearch. Because of the history of the project, some data we have does not
+conform to the requirements we now set and does not have the fields we describe
+as required, and there are some fields that store a very different data piece
+than what the users might expect[^1].
+
+The main goal of this project is to establish a comprehensive baseline of our
+media data: what information we currently have about each media item and what we
+expect to have. Our database currently holds more than 700 million rows of media
+data. The same data is in our API database and Elasticsearch, but the data
+properties there are slightly different. Due to the project's history, some data
+in the catalog does not meet our current requirements and lacks necessary
+fields, while other fields store data that doesn't align with user
+expectations[^1].
+
+By documenting all data properties, we can ensure that we have a complete
+understanding of the data. This documentation will improve the API user and the
+Openverse maintainer experience. It will also be the basis for data
+normalization which will lead to faster searches and a better experience for
+catalog and API contributors, as well as API users.
+
+### Success criteria
+
+The project will be considered successful when the document described below is
+created, reviewed by at least two maintainers, and published on the Openverse
+documentation site.
+
+### Implementors
+
+@rwidom - creator of the
+[image table](https://docs.google.com/spreadsheets/d/1gaVsvFnsYby2iwzRm0Ta9IPhvNixCpQdHByGLo1r_pg/edit?usp=sharing)
+("image table ddl" tab)
+
+@obulat - main implementor @openverse-team - reviewers
+
+### Implementation details
+
+As this is mainly a documentation project, its implementation plan can be
+written here, inside the project plan. (I think splitting it will unnecessarily
+slow down a smallish project. Please, correct me if you think I'm wrong) This
+project will create a table with all the media properties that we store or
+compute on all levels of the stack (Catalog, API database, API elasticsearch,
+Frontend). This table will serve as a basis for any work on data normalization
+work. As such, it will be a work in progress because the links to the code and
+the details about the properties should change as we progress through the data
+normalization project. We can publish it on the monorepo documentation site or
+the Openverse blog on Make WordPress, with the admonition that the table is a
+living document and will be updated.
```

Comment:
I don't know if it makes sense for this proposal, but I can imagine a world in which the provider data ingester has a required method for each of the required fields, and the spreadsheet is populated by the doc strings for those methods, and then there is some CI like what we have for the DAG doc strings. It would be different for SQL based providers, but those methods could return part of a SQL statement. It would be a big refactor to get from our world to that world, but if the goal is to move data normalization upstream, we might be able to get there... ???

Reply:
That could work for the catalog end perhaps! I was mostly thinking of a script that would go check the DDLs for our tables and ensure that each field is represented in the table (and vice versa). Just something very simple and straightforward for the cope of this effort. We could get more complicated down the road, but even a check to make sure that the [catalog DDL fields](https://github.com/WordPress/openverse-catalog/blob/b173d6e2dbf72b41728baed18ee8334d862b6be8/docker/local_postgres/0003_openledger_image_schema.sql#L16-L42) were present in the documentation and same for the [API model](https://github.com/WordPress/openverse/blob/main/api/catalog/api/models/image.py#L43) (though the latter would require something to actually load the `Image` model and access properties rather than just read lines in a file, which would be sufficient on the catalog end... 🤔).